### PR TITLE
Switch context feature. 

### PR DESCRIPTION
**Pull request by rafarui**
#### Please provide details about this Pull Request (why the change is being made, what this commit will do):
New feature. This allows the user to change the context of the Carol instance. It means that it is possible to get data from the environment A being in environment B. This only works if the user has access to that environment. 

Context: 
When we are running a carol app, all the credentials come from environment variables. Credentials for that app. An App can have multiples data sources in different environments.

During runtime, i don't have access to the credentials for those data sources. Either I add to the code this credentials, or I get from somewhere. With Carol's organizations, we can do this switch using APIs.

----

I have two implementations for that. 
1. channging the context from B to A. 
```python
carol = Carol('B', 'teste', auth=PwdAuth('email@totvs.com.br', 'pass'),  )
carol.switch_environment('A')
#now carol is instance of Tenant A
# switching back
carol.switch_environment('B')
```
or using context manager, where the switch only happens inside the contect manager.

```python
carol = Carol('B', 'teste', auth=PwdAuth('email@totvs.com.br', 'pass'), )
with carol.switch_context('onboarding') as A:
    .... #in tenant A
#back to tenant B
````

For the `PwdAuth` this does not make much sense. Since the password is the same for the same organization, but the ApiKey is different. So, I can do 

```python
carol = Carol('B', 'teste', auth=ApiKeyAuth(key),)
carol.switch_environment('A')
```
   
or

```python
carol = Carol('B', 'teste', auth=ApiKeyAuth(key),)
with carol.switch_context('onboarding') as A:
    .... #in tenant A
#back to tenant B
```

I will not support switch organizations until a case is presented. There are some caveats for this kind of change.

### CODE REVIEW

File: pycarol/organization.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+
+
+
+
+
+
+
+
+
```
Comment (gzamboni) [inline pycarol/organization.py]: Lots of blank lines?

File: pycarol/organization.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+class Organization:
+
```
Comment (gzamboni) [inline pycarol/organization.py]: no docstrings?

File: pycarol/organization.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    def __init__(self, carol):
+        self.carol = carol
+
+    def get_organization_info(self, organization):
```
Comment (gzamboni) [inline pycarol/organization.py]: If this is a class... where are the properties for an Organization object? 
Or
this is a helper module... so... no need to add those functions as class methods...

File: pycarol/organization.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    def get_environment_info(self, environment):
+        """
+        Get environment information.
+
+        :param environment: `str`
+            Tenant name
+        :return:
+            dict with the information about the environment.
+        """
+        return self.carol.call_api(f'v2/tenants/domain/{environment}',
```
Comment (gzamboni) [inline pycarol/organization.py]: what if environment == None or ""

File: pycarol/organization.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+                                   auth=False, status_forcelist=[], retries=0)
+
+    def all_tenants(self, org_id=None, org_name=None):
+
+        return self.carol.call_api(f'v1/organizations/{org_id}/allTenants',
```
Comment (gzamboni) [inline pycarol/organization.py]: what if org_id or org_name  == None or ""

File: pycarol/organization.py (hunk 3 above)
Comment (gzamboni) [inline pycarol/organization.py]: what if organization == None or ""

File: pycarol/organization.py (hunk 1 above)
Comment (rafarui) [reply pycarol/organization.py]: fixed,

File: pycarol/organization.py (hunk 4 above)
Comment (rafarui) [reply pycarol/organization.py]: actually, this should not be there. I still dont know if it will be possible to do this.

**Comment by rafarui** [comment]
> Reading this... what is the difference between switch the context inside the Carol object or have another object logged in a different context?
> 
> Would it be better to have 2 objects at the same time in different contexts?

When we are running a carol app, all the credentials come from environment variables. Credentials for that app. An App can have multiples data sources in different environments. 

During runtime, i don't have access to the credentials for those data sources.  Either I add to the code this credentials,  or I get from somewhere. With Carol's organizations, we can do this switch using APIs.

**Comment by gzamboni** [comment]
> #### Please provide details about this Pull Request (why the change is being made, what this commit will do):
> New feature. This allows the user to change the context of the Carol instance. It means that it is possible to get data from the environment A being in environment B. This only works if the user has access to that environment.
> 
> I have two implementations for that.
> 
> 1. channging the context from B to A.
> 
> ```python
> carol = Carol('B', 'teste', auth=PwdAuth('email@totvs.com.br', 'pass'),  )
> carol.switch_environment('A')
> #now carol is instance of Tenant A
> # switching back
> carol.switch_environment('B')
> ```
> 
> or using context manager, where the switch only happens inside the contect manager.
> 
> ```python
> carol = Carol('B', 'teste', auth=PwdAuth('email@totvs.com.br', 'pass'), )
> with carol.switch_context('onboarding') as A:
>     .... #in tenant A
> #back to tenant B
> ```



> > Reading this... what is the difference between switch the context inside the Carol object or have another object logged in a different context?
> > Would it be better to have 2 objects at the same time in different contexts?
> 
> When we are running a carol app, all the credentials come from environment variables. Credentials for that app. An App can have multiples data sources in different environments.
> 
> During runtime, i don't have access to the credentials for those data sources. Either I add to the code this credentials, or I get from somewhere. With Carol's organizations, we can do this switch using APIs.
> 
> PS: I will add this to the main comment in the PR. Since this is still a draft to get your opinions.
> I need to think about how to handle the switch of organizations too.

so... just for understanding...

what is the difference between:
```python
carol = Carol('B', 'teste', auth=PwdAuth('email@totvs.com.br', 'pass'),  )
carol.switch_environment('A')
#now carol is instance of Tenant A
# switching back
carol.switch_environment('B')
```
and 
```python
carol_A = Carol('A', 'teste', auth=PwdAuth('email@totvs.com.br', 'pass'),  )
carol_B = Carol('B', 'teste', auth=PwdAuth('email@totvs.com.br', 'pass'),  )
```

**Comment by rafarui** [comment]
> > #### Please provide details about this Pull Request (why the change is being made, what this commit will do):
> > New feature. This allows the user to change the context of the Carol instance. It means that it is possible to get data from the environment A being in environment B. This only works if the user has access to that environment.
> > I have two implementations for that.
> > 
> > 1. channging the context from B to A.
> > 
> > ```python
> > carol = Carol('B', 'teste', auth=PwdAuth('email@totvs.com.br', 'pass'),  )
> > carol.switch_environment('A')
> > #now carol is instance of Tenant A
> > # switching back
> > carol.switch_environment('B')
> > ```
> > 
> > 
> > or using context manager, where the switch only happens inside the contect manager.
> > ```python
> > carol = Carol('B', 'teste', auth=PwdAuth('email@totvs.com.br', 'pass'), )
> > with carol.switch_context('onboarding') as A:
> >     .... #in tenant A
> > #back to tenant B
> > ```
> 
> > > Reading this... what is the difference between switch the context inside the Carol object or have another object logged in a different context?
> > > Would it be better to have 2 objects at the same time in different contexts?
> > 
> > 
> > When we are running a carol app, all the credentials come from environment variables. Credentials for that app. An App can have multiples data sources in different environments.
> > During runtime, i don't have access to the credentials for those data sources. Either I add to the code this credentials, or I get from somewhere. With Carol's organizations, we can do this switch using APIs.
> > PS: I will add this to the main comment in the PR. Since this is still a draft to get your opinions.
> > I need to think about how to handle the switch of organizations too.
> 
> so... just for understanding...
> 
> what is the difference between:
> 
> ```python
> carol = Carol('B', 'teste', auth=PwdAuth('email@totvs.com.br', 'pass'),  )
> carol.switch_environment('A')
> #now carol is instance of Tenant A
> # switching back
> carol.switch_environment('B')
> ```
> 
> and
> 
> ```python
> carol_A = Carol('A', 'teste', auth=PwdAuth('email@totvs.com.br', 'pass'),  )
> carol_B = Carol('B', 'teste', auth=PwdAuth('email@totvs.com.br', 'pass'),  )
> ```

virtually none since you are using user/password that is always the same inside an organization. 

The problem is that you dont have acess to `PwdAuth` when running inside a carol app. 
Only the API token for tenant A. I mean, this is not true. 

```python 
carol_A = Carol('A', 'teste', auth=ApiKeyAuth(KEY_TENANT_A)),  )
carol_B = Carol('B', 'teste', auth=ApiKeyAuth(KEY_TENANT_A)),  )
```
`ApiKeyAuth` for tenant A is not the same that `ApiKeyAuth` for tenant B. like password. 

maybe a better example. 

```pytohn
carol = Carol('B', 'teste', auth=ApiKeyAuth(key),)
carol.switch_environment('B')
```

**Comment by gzamboni** [comment]
> > > #### Please provide details about this Pull Request (why the change is being made, what this commit will do):
> > > New feature. This allows the user to change the context of the Carol instance. It means that it is possible to get data from the environment A being in environment B. This only works if the user has access to that environment.
> > > I have two implementations for that.
> > > 
> > > 1. channging the context from B to A.
> > > 
> > > ```python
> > > carol = Carol('B', 'teste', auth=PwdAuth('email@totvs.com.br', 'pass'),  )
> > > carol.switch_environment('A')
> > > #now carol is instance of Tenant A
> > > # switching back
> > > carol.switch_environment('B')
> > > ```
> > > 
> > > 
> > > or using context manager, where the switch only happens inside the contect manager.
> > > ```python
> > > carol = Carol('B', 'teste', auth=PwdAuth('email@totvs.com.br', 'pass'), )
> > > with carol.switch_context('onboarding') as A:
> > >     .... #in tenant A
> > > #back to tenant B
> > > ```
> > 
> > 
> > > > Reading this... what is the difference between switch the context inside the Carol object or have another object logged in a different context?
> > > > Would it be better to have 2 objects at the same time in different contexts?
> > > 
> > > 
> > > When we are running a carol app, all the credentials come from environment variables. Credentials for that app. An App can have multiples data sources in different environments.
> > > During runtime, i don't have access to the credentials for those data sources. Either I add to the code this credentials, or I get from somewhere. With Carol's organizations, we can do this switch using APIs.
> > > PS: I will add this to the main comment in the PR. Since this is still a draft to get your opinions.
> > > I need to think about how to handle the switch of organizations too.
> > 
> > 
> > so... just for understanding...
> > what is the difference between:
> > ```python
> > carol = Carol('B', 'teste', auth=PwdAuth('email@totvs.com.br', 'pass'),  )
> > carol.switch_environment('A')
> > #now carol is instance of Tenant A
> > # switching back
> > carol.switch_environment('B')
> > ```
> > 
> > 
> > and
> > ```python
> > carol_A = Carol('A', 'teste', auth=PwdAuth('email@totvs.com.br', 'pass'),  )
> > carol_B = Carol('B', 'teste', auth=PwdAuth('email@totvs.com.br', 'pass'),  )
> > ```
> 
> virtually none since you are using user/password that is always the same inside an organization.
> 
> The problem is that you dont have acess to `PwdAuth` when running inside a carol app.
> Only the API token for tenant A. I mean, this is not true.
> 
> ```python
> carol_A = Carol('A', 'teste', auth=ApiKeyAuth(KEY_TENANT_A)),  )
> carol_B = Carol('B', 'teste', auth=ApiKeyAuth(KEY_TENANT_A)),  )
> ```
> 
> `ApiKeyAuth` for tenant A is not the same that `ApiKeyAuth` for tenant B. like password.
> 
> maybe a better example.
> 
> ```
> carol = Carol('B', 'teste', auth=ApiKeyAuth(key),)
> carol.switch_environment('B')
> ```

I see.. the root problem is... the Token belongs to the Tenant... not to the User... now it makes sense.

**Comment by rafarui** [comment]
Yes!!. I will change the ex. to use the APIToken

File: pycarol/organization.py (hunk 3 above)
Comment (rafarui) [reply pycarol/organization.py]: I might need this to implement the switch for organizations. 
Still trying to solve this

**Pull request merged**